src/replication.c

```diff
@@ -197,27 +202,207 @@ void feedReplicationBacklogWithObject(robj *o) {
+        tail->refcount = 0;
+        tail->repl_offset = server.master_repl_offset - tail->used + 1;
+        tail->id = repl_block_id++;
+        memcpy(tail->buf, s, len);
+        listAddNodeTail(server.repl_buffer_blocks, tail);
+        server.repl_buffer_size += (usable_size + sizeof(listNode));
+        add_new_block = 1;
+        if (start_node == NULL) {
+            start_node = listLast(server.repl_buffer_blocks);
+            start_pos = 0;
+        }
+    }
+
+    /* For output buffer of replicas. */
+    listIter li;
+    listRewind(server.slaves,&li);
+    while((ln = listNext(&li))) {
+        client *slave = ln->value;
+        if (!canFeedReplicaReplBuffer(slave)) continue;
 
-    return 1;
+        /* Update shared replication buffer start position. */
+        if (slave->ref_repl_buf_node == NULL) {
+            slave->ref_repl_buf_node = start_node;
+            slave->ref_block_pos = start_pos;
+            /* Only increase the start block reference count. */
+            ((replBufBlock *)listNodeValue(start_node))->refcount++;
+            /* Use last not full block. */
+            if (used_last_block) {
+                slave->used_blocks_of_repl_buf++;
+                slave->used_size_of_repl_buf += used_last_block->size;
+            }
+        }
+
+        /* Incr used replication buffer size, check output buffer limit
+         * only when add new block. */
+        if (add_new_block) {
+            slave->used_blocks_of_repl_buf++;
+            slave->used_size_of_repl_buf += tail->size;
+            closeClientOnOutputBufferLimitReached(slave, 1);
+        }
+    }
+
+    /* For replication backlog */
+    if (server.repl_backlog->ref_repl_buf_node == NULL) {
+        server.repl_backlog->ref_repl_buf_node = start_node;
+        /* Only increase the start block reference count. */
+        ((replBufBlock *)listNodeValue(start_node))->refcount++;
+
+        /* Replication buffer must be empty before adding replication stream
+         * into replication backlog. */
+        serverAssert(used_last_block == NULL);
+    }
+    if (add_new_block) {
+        /* To make search offset from replication buffer blocks quickly
+         * when replicas ask partial resynchronization, we record one block
+         * every 64 blocks. */
+        server.repl_backlog->unrecorded_count++;
+        if (server.repl_backlog->unrecorded_count >= RECORD_ONE_EVERY_BLOCKS) {
+            uint64_t encoded_offset = htonu64(tail->repl_offset);
+            raxInsert(server.repl_backlog->blocks_index,
+                (unsigned char*)&encoded_offset, sizeof(uint64_t),
+                listLast(server.repl_buffer_blocks), NULL);
+            server.repl_backlog->unrecorded_count = 0;
+        }
+    }
+    /* Trim replication backlog. */
+    incrementalTrimReplicationBacklog(TRIM_REPL_BUF_BLOCKS_PER);
+}
+
+/* Free replication buffer blocks that are referenced by this client. */
+void freeSlaveReferencedReplBuffer(client *replica) {
```

Comment:
let's try to avoid the S word when adding new code.
```suggestion
void freeReplicaReferencedReplBuffer(client *replica) {
```